src/net/sourceforge/kolmafia/session/MallPriceManager.java

```diff
@@ -154,92 +154,81 @@ public static MallSearchRequest newMallSearchRequest(String category, String tie
+      if (search != null && MallPriceManager.removeShopPurchaseRequest(itemId, shopId, search)) {
+        if (search.size() == 0) {
+          MallPriceManager.mallSearches.remove(itemId);
+        }
       }
+      return;
+    }
 
-      Iterator<PurchaseRequest> i2 = search.iterator();
-      while (i2.hasNext()) {
-        PurchaseRequest purchase = i2.next();
-        if (purchase instanceof MallPurchaseRequest) {
-          MallPurchaseRequest mallPurchase = (MallPurchaseRequest) purchase;
-          if (shopId == mallPurchase.getShopId()) {
-            i2.remove();
-            MallPriceManager.updateMallPrice(ItemPool.get(itemId), search);
-            if (itemId != -1) {
-              return;
-            }
-            break;
-          }
+    // Remove shop from search results for all items
+    Iterator<Entry<Integer, List<PurchaseRequest>>> i =
+        MallPriceManager.mallSearches.entrySet().iterator();
+    while (i.hasNext()) {
+      Entry<Integer, List<PurchaseRequest>> entry = i.next();
+      int key = entry.getKey();
+      List<PurchaseRequest> search = entry.getValue();
+      if (MallPriceManager.removeShopPurchaseRequest(key, shopId, search)) {
+        if (search.size() == 0) {
+          i.remove();
         }
       }
     }
   }
 
   public static final void flushCache(final int itemId) {
-    Iterator<List<PurchaseRequest>> i = MallPriceManager.mallSearches.values().iterator();
-    while (i.hasNext()) {
-      List<PurchaseRequest> search = i.next();
-      // Always remove empty searches
-      if (search == null || search.size() == 0) {
-        i.remove();
-        continue;
-      }
-      int id = search.get(0).getItemId();
-      if (itemId == id) {
-        i.remove();
-        MallPriceManager.updateMallPrice(ItemPool.get(itemId), search);
-        return;
-      }
-      break;
+    List<PurchaseRequest> search = MallPriceManager.mallSearches.get(itemId);
+    if (search != null) {
+      MallPriceManager.mallSearches.remove(itemId);
+      // *** Previous code had this. Really?
+      // MallPriceManager.updateMallPrice(ItemPool.get(itemId), search);
+      MallPriceManager.mallPrices.put(itemId, 0);
     }
   }
 
-  public static final void flushCache() {
-    long t0, t1;
-    t1 = MallPriceManager.currentTimeMillis();
-    t0 = t1 - MALL_SEARCH_FRESHNESS * 1000;
-
-    Iterator<List<PurchaseRequest>> i = MallPriceManager.mallSearches.values().iterator();
-    while (i.hasNext()) {
-      List<PurchaseRequest> search = i.next();
-      if (search == null || search.size() == 0) {
-        i.remove();
-        continue;
-      }
-      long t = search.get(0).getTimestamp();
-      if (t < t0 || t > t1) {
-        i.remove();
-        continue;
-      }
-      break;
+  public static boolean searchIsTooOld(List<PurchaseRequest> search) {
+    if (search == null || search.size() == 0) {
+      return true;
     }
+    long t1 = MallPriceManager.currentTimeMillis();
```

Comment:
While I realise this was a pre-existing problem, can we have better names than `t0` and `t1`? `freshnessLimit` and `now`?

Also the search is too old if the timestamp is after the current time? Why is that? (Again a pre-existing issue though).

Reply:
Done.